From: Hendrik Brueckner <brueckner@redhat.com>
Date: Wed, 20 Jan 2010 13:38:10 -0500
Subject: [scsi] scsi_transport_fc: Allow LLD to reset FC BSG timeout
Message-id: <1263994692-1095-2-git-send-email-brueckner@redhat.com>
Patchwork-id: 22675
O-Subject: [RHEL6 PATCH 1/3] [scsi] scsi_transport_fc: Allow LLD to reset FC BSG
	timeout
Bugzilla: 556918
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Mike Christie <mchristi@redhat.com>

Description
-----------
Kernel panic might occur after a FC BSG requests is aborted while
the corresponding zfcp request is still being processed.

The problem is caused by the zfcp hardware which cannot about a
currently pending CT or ELS request.

The solution is to postpone the triggered timeout request within
the FC layer, since there is nothing zfcp can do to stop the
request at this point.

Bugzilla
--------
BZ 556918
https://bugzilla.redhat.com/show_bug.cgi?id=556918

Upstream status of the patch
----------------------------
The patch has been accepted in the scsi-rc-fixes tree and
will be pulled upstream for kernel version 2.6.33.
http://git.kernel.org/?p=linux/kernel/git/jejb/scsi-rc-fixes-2.6.git;a=commit;h=b8f08645f80a0c93246f3539ce53d3cae6a91fa0

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/drivers/scsi/scsi_transport_fc.c b/drivers/scsi/scsi_transport_fc.c
index 39e0e12..ee091b8 100644
--- a/drivers/scsi/scsi_transport_fc.c
+++ b/drivers/scsi/scsi_transport_fc.c
@@ -3511,7 +3511,10 @@ fc_bsg_job_timeout(struct request *req)
 	if (!done && i->f->bsg_timeout) {
 		/* call LLDD to abort the i/o as it has timed out */
 		err = i->f->bsg_timeout(job);
-		if (err)
+		if (err == -EAGAIN) {
+			job->ref_cnt--;
+			return BLK_EH_RESET_TIMER;
+		} else if (err)
 			printk(KERN_ERR "ERROR: FC BSG request timeout - LLD "
 				"abort failed with status %d\n", err);
 	}
